fix(core): Flush instance stopped event immediately

tomi opened the pull request:
## Summary

1. fix(core): Flush instance stopped event immediately

Rudderstack flushes events after 20 events have been batched or
10 seconds have elapsed since the last flush. Since we were
waiting for the 'User instance stopped' event to be sent before
calling flush, we might have waited up to 10 seconds before
the event got sent. Instead call the flush right away to send the
stopped and any other remaining events immediately.

2.  refactor(core): Make telemetry track & identify synchronous

We are not actually waiting for the track & identify events anywhere,
so it doesn't make sense to have the method async.


## Review / Merge checklist

- [ ] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) <!--
   **Remember, the title automatically goes into the changelog.
   Use `(no-changelog)` otherwise.**
-->
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [ ] Tests included. <!--
   A bug is not considered fixed, unless a test is added to prevent it from happening again.
   A feature is not complete without tests.
-->
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

tomi committed:
fix(core): Flush instance stopped event immediately

Rudderstack flushes events after 20 events have been batched or
10 seconds have elapsed since the last flush. Since we were
waiting for the 'User instance stopped' event to be sent before
calling flush, we might have waited up to 10 seconds before
the event got sent. Instead call the flush right away to send the
stopped and any other remaining events immediately.

netroy commented on packages/cli/src/telemetry/index.ts:
```diff
@@ -171,8 +171,13 @@ export class Telemetry {
-		void Promise.all([this.postHog.stop(), this.rudderStack?.flush()]);
+
+		// Send the stopped event and make sure it gets flushed right away
+		await Promise.all([
```
Promise.all does not guarantee a sequential execution. the track call might happen before the flush.

tomi replied:
No it doesn't, and here it doesn't need to. The track call pushes to the queue synchronously, and returns a promise that resolves when the event is sent. This happens synchronously before the flush call is executed. The flush then sends the event and any other events in the queue. Very confusing yes.

I could have written it like this:

```ts
void this.track()

await Promise.all([this.postHog.stop(), this.rudderStack?.flush()]
```

But I noticed that RS's `flush` only [sends at max 20 events at a time](https://github.com/rudderlabs/rudder-sdk-node/blob/master/src/index.js#L625) (unless `flushAt` is [changed](https://www.rudderstack.com/docs/sources/event-streams/sdks/rudderstack-node-sdk/#sdk-initialization-options)). So theoretically if there are more than 20 events in the queue it might not get sent with the first flush. Altho looking at the code again it might be that that scenario never happens.

I think we could actually simplify by making the `track` function entirely synchronous, as I couldn't find any cases where it would actually make sense to wait until the event has been sent. And if that is actually the case, one can always call `flush`. WDYT?

netroy replied:
> I think we could actually simplify by making the `track` function entirely synchronous

Yes please 🙏🏽

tomi replied:
@netroy refactor all the `track` and `identify` calls to be synchronous. Please have a look 🙏

tomi committed:
refactor(core): Make telemetry track & identify synchronous

We are not actually waiting for the track & identify events anywhere,
so it doesn't make sense to have the method async.

janober commented:
Got released with [`n8n@1.53.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.53.0)
